lib/spack/spack/build_systems/msbuild.py

```diff
@@ -0,0 +1,123 @@
+        |                                               | install phase       |
+        +-----------------------------------------------+---------------------+
+        | :py:meth:`~.MSBuildBuilder.build_directory`   | Directory where the |
+        |                                               | project sln/vcxproj |
+        |                                               | is located          |
+        +-----------------------------------------------+---------------------+
+    """
+
+    phases = ("build", "install")
+
+    #: Targets for ``make`` during the :py:meth:`~.MakefileBuilder.build` phase
+    build_targets: List[str] = []
+    #: Targets for ``make`` during the :py:meth:`~.MakefileBuilder.install` phase
+    install_targets: List[str] = ["INSTALL"]
+
+    @property
+    def build_directory(self):
+        """Return the directory containing the MSBuild solution or vcxproj."""
+        return self.pkg.stage.source_path
+
+    @property
+    def toolchain_version(self):
+        """Return currently targeted version of MSVC toolchain
+        Override this method to select a specific version of the toolchain or change
+        selection heuristics.
+        Default is whatever version of msvc has been selected by concretization"""
+        return self.compiler.msvc_version
+
+    @property
+    def std_msbuild_args(self):
+        """Return common msbuild cl arguments, for now just toolchain"""
+        return [self.define("PlatformToolset", self.toolchain_version)]
+
+    def define_targets(self, *targets):
+        return "/target:" + ";".join(targets) if targets else ""
+
+    def define(self, msbuild_arg, value):
+        return "/p:{}={}".format(msbuild_arg, value)
+
+    def msbuild_args(self):
+        """Define build arguments to MSbuild. This is an empty list by default.
+        Individual packages should override to specify MSBuild args to command line
+        PlatformToolset is already defined an can be controlled via the `toolchain_version`
+        property"""
+        return []
+
+    def msbuild_install_args(self):
+        """Define install arguments to MSBuild outside of the INSTALL target. This is the same
+        as `msbuild_args` by default."""
+        return self.msbuild_args()
+
+    def build(self, pkg, spec, prefix):
+        """Run "msbuild" on the build targets specified by the builder."""
+        with fs.working_dir(self.build_directory):
+            inspect.getmodule(self.pkg).msbuild(
+                *self.msbuild_args(), self.define_targets(*self.build_targets)
+            )
+
+    def install(self, pkg, spec, prefix):
+        """Run "msbuild" on the install targets specified by the builder.
+        This is INSTALL by default"""
+        with fs.working_dir(self.build_directory):
+            inspect.getmodule(self.pkg).msbuild(
+                *self.msbuild_install_args(), self.define_targets(*self.install_targets)
+            )
+
+    # def setup_dependent_package(self, m, dep_spec):
```

Comment:
If build systems other than `msbuild` use these, then I can see the appeal of defining them in `build_environment` (similar to why `make` is defined there).

Reply:
Only msbuild/nmake uses these, although there are currently generic classes using them, but those should be refactored. However, when I switched to using it this way, all nmake and msbuild `build`  builder phases fail because `nmake` or `msbuild` respectively are not defined, and it seems like this method is not invoked before the build stage. It seems as though other build tools either have packages to represent them, allowing for these tools to be defined when the package is loaded, or are defined in the `build_environment` module. According to the docs, `setup_dependent_packages` is only invoked before `install`, which makes it unsuitable for use in this case, as we need it during the `build` phase. If this is not the case, then something is broken and the docs need updating. If this is the case, can we implement something that does run before `build` or `configure`? Or maybe this is more configurable than I think and I'm not seeing it in the build system docs.

Reply:
`setup_dependent_packages` (and all `*setup*` functions) should run before all phases and it is indeed a bug if they are not run before the `build` phase - can you mention the error message here?

Reply:
The error:
```
Traceback (most recent call last):
  File "C:\spack_win\spack\lib\spack\spack\build_environment.py", line 1053, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "C:\spack_win\spack\lib\spack\spack\installer.py", line 2081, in build_process
    return installer.run()
  File "C:\spack_win\spack\lib\spack\spack\installer.py", line 1945, in run
    self._real_install()
  File "C:\spack_win\spack\lib\spack\spack\installer.py", line 2041, in _real_install
    phase_fn.execute()
  File "C:\spack_win\spack\lib\spack\spack\builder.py", line 420, in execute
    self.phase_fn(pkg, pkg.spec, pkg.prefix)
  File "C:\spack_win\spack\var\spack\repos\builtin\packages\winlibiconv\package.py", line 37, in build
NameError: name 'nmake' is not defined
```

When I include a line that, if the `setup_dependent_package` is invoked will exit with failure, nothing happens and execution continues as normal. Perhaps I'm doing something wrong because even if I remove the contents of the build method, the install method fails the same way, so either I've done something wrong, or this method is never being invoked. Considering I have a `raise` and an `exit(1)` in that stanza and nothing happens, I feel like it might never be invoked.

Reply:
Whoops sorry it just occurred to me: this should go into `setup_package`, not `setup_dependent_package`: if you have A->B, then setup_dependent_package is run for B when building A, but not for B when building B. Does that make sense?

Reply:
It does make sense, and I did think it was a bit of a bizarre phrasing for the function name. I just assumed it was for "packages dependent on the builder". That being said, switching to `setup_package` changes nothing, nor can I find a reference to that method being invoked anywhere except for a `build_system` context. It's a method in `build_environment` that invokes some environmental information about the package, but doesn't seem to be invoked by anything that would call our packages `setup_package` method, at least according to a grep of the codebase. 

From what I can tell, we rely on `setup_dependent_package` for almost every build system but autotools/makefiles and maven, as they all have representative packages. Autotools pulls from `build_environment` as we know, and maven runs a `which` during the install phase. I'm happy to add stub packages for nmake and msbuild that don't install, but do allow for external detection but being required to do that feels like a bug with the way things are currently structured.

Maybe we should sync about this later today?

Reply:
@alalazo it turns out that for now we need to keep defining `cscript`/`msbuild` in `build_environment`:

* We're able to customize CMake because packages which use it depend on the `cmake` Spack package, which does `setup_dependent_package`
* There is currently no Spack package which provides `msbuild`/`cscript`
* We don't have a way for build systems to indicate that they should modify the package module, i.e. there is no `setup_package` analog to `setup_dependent_package` (to avoid confusion, there *is* a method called `setup_package` but it is not used for this purpose)

For now I think the path forward is to:

* Define `msbuild`/`cscript` in `build_environment`
* Add a TODO message there
* In a follow-up PR, define a dependency package, or update the API to allow build systems to modify the module

Does that sound reasonable?